### ISSUE TITLE
Device safety: enforce GPU/CPU boundaries throughout the LF2I pipeline

### ISSUE THREAD

**Issue by JamesCarzon** [lee-group-cmu/lf2i#32]
## Background

Several places in the pipeline call `.numpy()` directly on `torch.Tensor` objects that may reside on CUDA, which raises a `RuntimeError` at runtime. Additionally, `joblib.Parallel` (used in `ACORE`, `BFF`, and `Waldo` evaluation loops) defaults to process-based parallelism, which cannot share CUDA contexts across forked processes — leading to silent corruption on Linux (fork) or a hard crash on spawn.

**Invariant to enforce:**
- **GPU allowed for:** NN calibration model training; test statistic estimator training and inference.
- **CPU required for:** all preprocessing helpers, Neyman inversion, diagnostics, and plotting.
- **At every GPU→CPU boundary:** use `to_np_if_torch()` (from `lf2i.utils.miscellanea`) rather than bare `.numpy()`.

> **Note on file paths:** an earlier draft of this issue referenced `src/lf2i/critical_values/quantile_regression.py`, `src/lf2i/critical_values/nn_qr_algorithm.py`, and `src/lf2i/diagnostics/diagnostics.py`. The correct paths are `src/lf2i/calibration/critical_values.py`, `src/lf2i/calibration/torch_utils.py`, and `src/lf2i/diagnostics/coverage_probability.py`.

---

## Step 1 — Fix all unsafe `.numpy()` calls

All fixes are mechanical: replace bare `tensor.numpy()` with `to_np_if_torch(tensor)` (add `from lf2i.utils.miscellanea import to_np_if_torch` where not already imported), or inline `tensor.cpu().detach().numpy()`. No logic changes.

**`src/lf2i/utils/calibration_diagnostics_inputs.py`**

- [x] `preprocess_train_quantile_regression`, lines 38–39 — `test_statistics.numpy()` / `parameters.numpy()`
- [x] `preprocess_predict_quantile_regression`, line 65 — `parameters.numpy()`
- [x] `preprocess_fit_p_values`, line 83 — `inp.numpy()`
- [x] `preprocess_predict_p_values`, lines 112–113 — `test_stats.numpy()` / `poi.numpy()`
- [x] `preprocess_diagnostics`, lines 166–169 — three bare `.numpy()` calls on `indicators`, `parameters`, `new_parameters`
- [x] `preprocess_indicators_prediction`, lines 226–228 — `parameters.numpy()` / `samples.numpy()`

**`src/lf2i/utils/odds_inputs.py`**

- [x] `preprocess_odds_estimation`, lines 31 and 33 — `parameters.numpy()` / `samples.numpy()` (sklearn/XGBoost branch)
- [x] `preprocess_for_odds_cv`, lines 280–282 — `parameters.numpy()` / `samples.numpy()` (numpy branch)
- [x] `preprocess_for_odds_cs`, lines 370–371 — `parameter_grid.numpy()` / `samples.numpy()`

**`src/lf2i/utils/waldo_inputs.py`**

- [x] `preprocess_waldo_estimation`, lines 31–33 — `parameters.numpy()` / `samples.numpy()` (sklearn/XGBoost branch)
- [x] `preprocess_waldo_computation`, line 88 — `parameters.numpy()`

**`src/lf2i/calibration/p_values.py`**

- [x] `augment_calibration_set`, lines 126–128 — `test_statistics.numpy()` / `poi.numpy()`

---

## Step 2 — Fix Waldo posterior `.numpy()` crash

- [x] `src/lf2i/test_statistics/waldo.py`, line 207:

  ```python
  # before
  posterior_samples = self.estimator.sample(...).numpy()

  # after
  posterior_samples = self.estimator.sample(...).cpu().numpy()
  ```

  If the SBI posterior estimator runs on CUDA, bare `.numpy()` raises `RuntimeError`. Calling `.cpu()` first is safe regardless of device.

---

## Step 3 — Resolve diagnostics TODO

- [x] `src/lf2i/diagnostics/coverage_probability.py`, line 165: remove the `# TODO: convert torch Tensors into numpy arrays` comment. The conversion should be enforced inside `preprocess_indicators_lf2i` in `calibration_diagnostics_inputs.py` — add `to_np_if_torch()` calls on `test_statistics`, `critical_values`, `p_values`, and `parameters` before the reshape operations. This makes the function safe to call with either numpy arrays or CUDA tensors.

---

## Step 4 — Re-enable GPU for NN calibration + thread-based parallelism for GPU estimators

Steps 4a–4c must be committed together: re-enabling CUDA on the calibration NN is only safe once the `Parallel` loops in test statistic evaluation switch to threads for GPU models (process-based `fork`/`spawn` cannot share a CUDA context).

### 4a — Add `_estimator_on_gpu()` helper

- [x] Add to `src/lf2i/utils/miscellanea.py`:

  ```python
  def _estimator_on_gpu(estimator) -> bool:
      """Return True if estimator is a PyTorch nn.Module with parameters on CUDA."""
      if isinstance(estimator, torch.nn.Module):
          try:
              return next(estimator.parameters()).device.type == 'cuda'
          except StopIteration:
              return False
      if hasattr(estimator, 'model') and isinstance(estimator.model, torch.nn.Module):
          try:
              return next(estimator.model.parameters()).device.type == 'cuda'
          except StopIteration:
              return False
      return False
  ```

  PyTorch releases the GIL during C++/CUDA kernel execution, so threads safely share a GPU model and its CUDA context. GPU kernel launches remain serialized by CUDA internally, but CPU-side work (preprocessing, result gathering) runs in parallel — which is the dominant cost at transformer scale.

### 4b — Switch `Parallel` calls to threading when estimator is on GPU (1d7a13039eda60477f03db0cd902a353edc86644)

At each `Parallel(n_jobs=self.n_jobs)(...)` call in the files below, add `prefer='threading' if _estimator_on_gpu(self.estimator) else 'loky'`. Import `_estimator_on_gpu` from `lf2i.utils.miscellanea`.

**Note:** Some keywords in Parallel are not up-to-date -- use "threads" and "processes" in `prefer` argument

- [x] `src/lf2i/test_statistics/acore.py` — all `Parallel(...)` calls in `_compute_for_critical_values`, `_compute_for_confidence_sets`, `_compute_for_diagnostics`, and `_compute_restricted_mle_for_confidence_sets`
- [x] `src/lf2i/test_statistics/bff.py` — all `Parallel(...)` calls in `_compute_for_critical_values` and `_compute_for_confidence_sets`
- [x] `src/lf2i/test_statistics/waldo.py` — the `Parallel(...)` call in `evaluate()` (posterior branch)

### 4c — Re-enable CUDA for NN calibration model training

- [x] `src/lf2i/calibration/critical_values.py`, line 119:

  ```python
  # before
  device='cpu',  # "cuda" if torch.cuda.is_available() else 'cpu',

  # after
  device="cuda" if torch.cuda.is_available() else "cpu",
  ```

  `LearnerRegression.predict()` already returns `.cpu().detach()` and `Learner.fit()` already does `batch_loss.cpu().detach().numpy()`, so the CPU boundary at the calibration output is already managed.

---

## Step 5 — Tests

- [ ] Test that `LearnerRegression.predict()` returns a CPU array when `device='cuda'` (mark with `pytest.mark.skipif(not torch.cuda.is_available(), reason='no GPU')`)
- [ ] Test `Waldo.evaluate()` with a mock posterior whose `.sample()` returns a CUDA tensor; assert the output is a CPU `np.ndarray`
- [ ] For each `preprocess_*` function edited in Step 1: parametrize with a CUDA tensor input, assert the output is a CPU `np.ndarray` (skip if no GPU)

---

### Batch `predict_proba` inside the ACORE denominator optimizer

`_maximize_log_odds` in `acore.py` calls `predict_proba` once per optimizer step *per observation*. With a GPU model this is inefficient: each call incurs kernel launch overhead for a single sample. The fix is to accumulate all observations' `(parameter, sample)` inputs for a given optimizer step into one batch tensor and issue a single `predict_proba` call, then split results per observation. The same pattern applies to the BFF integration loop. This is a moderate refactor that significantly reduces GPU round-trips when running large calibration or diagnostic sweeps.

### Full vectorization of `evaluate()`

Remove the per-observation loop entirely. `preprocess_for_odds_cs` and `preprocess_for_odds_cv` already construct the full stacked `(parameter, sample)` tensor — a single batched `predict_proba` call can handle all observations at once. The main prerequisite is vectorizing the denominator optimizer across observations, which depends on the batching work above.
